libafl_frida/src/alloc.rs

```diff
@@ -0,0 +1,408 @@
+                );
+            }
+            return;
+        };
+
+        if metadata.freed {
+            AsanErrors::get_mut().report_error(
+                AsanError::DoubleFree((ptr as usize, metadata.clone(), Backtrace::new())),
+                None,
+            );
+        }
+        let shadow_mapping_start = map_to_shadow!(self, ptr as usize);
+
+        metadata.freed = true;
+        if self.options.enable_asan_allocation_backtraces {
+            metadata.release_site_backtrace = Some(Backtrace::new_unresolved());
+        }
+
+        // poison the shadow memory for the allocation
+        Self::poison(shadow_mapping_start, metadata.size);
+    }
+
+    pub fn find_metadata(
+        &mut self,
+        ptr: usize,
+        hint_base: usize,
+    ) -> Option<&mut AllocationMetadata> {
+        let mut metadatas: Vec<&mut AllocationMetadata> = self.allocations.values_mut().collect();
+        metadatas.sort_by(|a, b| a.address.cmp(&b.address));
+        let mut offset_to_closest = i64::max_value();
+        let mut closest = None;
+        for metadata in metadatas {
+            let new_offset = if hint_base == metadata.address {
+                (ptr as i64 - metadata.address as i64).abs()
+            } else {
+                std::cmp::min(
+                    offset_to_closest,
+                    (ptr as i64 - metadata.address as i64).abs(),
+                )
+            };
+            if new_offset < offset_to_closest {
+                offset_to_closest = new_offset;
+                closest = Some(metadata);
+            }
+        }
+        closest
+    }
+
+    pub fn reset(&mut self) {
+        for (address, mut allocation) in self.allocations.drain() {
+            // First poison the memory.
+            Self::poison(map_to_shadow!(self, address), allocation.size);
+
+            // Reset the allocaiton metadata object
+            allocation.size = 0;
+            allocation.freed = false;
+            allocation.allocation_site_backtrace = None;
+            allocation.release_site_backtrace = None;
+
+            // Move the allocation from the allocations to the to-be-allocated queues
+            self.allocation_queue
+                .entry(allocation.actual_size)
+                .or_default()
+                .push(allocation);
+        }
+    }
+
+    pub fn get_usable_size(&self, ptr: *mut c_void) -> usize {
+        match self.allocations.get(&(ptr as usize)) {
+            Some(metadata) => metadata.size,
+            None => {
+                panic!(
+                    "Attempted to get_usable_size on a pointer ({:?}) which was not allocated!",
+                    ptr
+                );
+            }
+        }
+    }
+
+    fn unpoison(start: usize, size: usize) {
+        //println!("unpoisoning {:x} for {:x}", start, size / 8 + 1);
+        unsafe {
+            //println!("memset: {:?}", start as *mut c_void);
+            memset(start as *mut c_void, 0xff, size / 8);
```

Comment:
Just FYI you can use a normal rust loop here, the compiler will do the magic like memset etc

Reply:
Are you sure? Wanted to make sure I'm using the most efficient implementation...

Reply:
Yes

Reply:
Reference https://rust.godbolt.org/z/av31cd455

Reply:
Lol the code I pasted into the reference is wrong fwiw :D but it shows it's using memset also.